File: query_primer/query_primer__dount.cs
using static System.Console;
using static System.Linq.Enumerable;

internal class Program
{
    static void Main()
    {
        int[] hwn = ReadLine().Split().Select( int.Parse ).ToArray();
        int[,] a = new int[hwn[0], hwn[1]];
        int[,] s = new int[hwn[0] + 1, hwn[1] + 1];

        SetOut( new System.IO.StreamWriter( OpenStandardOutput() ) { AutoFlush = false } );
        foreach( int i in Range( 0, hwn[0] ) ) {
            int[] t = ReadLine().Split().Select( int.Parse ).ToArray();

            foreach( int j in Range( 0, hwn[1] ) ) {
                a[i, j] = t[j];
            }
        }
        foreach( int i in Range( 0, hwn[0] ) ) {
            s[i, 0] = 0;
            foreach( int j in Range( 0, hwn[1] ) ) {
                s[i + 1, j + 1] = s[i + 1, j] + s[i, j + 1] - s[i, j] + a[i, j];
            }
        }
        foreach( int _ in Range( 0, hwn[2] ) ) {
            int[] t = ReadLine().Split().Select( x => int.Parse( x ) - 1 ).ToArray();
            int m = t[2] / 2;
            int lt = s[t[0] - m, t[1] - m];
            int lb = s[t[0] + m + 1, t[1] - m];
            int rt = s[t[0] - m, t[1] + m + 1];
            int rb = s[t[0] + m + 1, t[1] + m + 1];
            int os = rb - rt - lb + lt;

            m = t[3] / 2;
            lt = s[t[0] - m, t[1] - m];
            lb = s[t[0] + m + 1, t[1] - m];
            rt = s[t[0] - m, t[1] + m + 1];
            rb = s[t[0] + m + 1, t[1] + m + 1];
            WriteLine( os - ( rb - rt - lb + lt ) );
        }
        Out.Flush();
        return;
    }
}

File: query_primer/query_primer__dount.py
h, w, n = list(map(int, input().split()))
a = [[0] * w for _ in range(h)]
s = [[0] * (w + 1) for _ in range(h + 1)]

for i in range(h):
    t = list(map(int, input().split()))

    for j in range(w):
        a[i][j] = t[j]
for i in range(h):
    for j in range(w):
        s[i + 1][j + 1] = s[i + 1][j] + s[i][j + 1] - s[i][j] + a[i][j]
for _ in range(n):
    t = list(map(int, input().split()))
    for i in range(len(t)):
        t[i] -= 1
    m = t[2] // 2
    lt = s[t[0] - m][t[1] - m]
    lb = s[t[0] + m + 1][t[1] - m]
    rt = s[t[0] - m][t[1] + m + 1]
    rb = s[t[0] + m + 1][t[1] + m + 1]
    os = rb - rt - lb + lt

    m = t[3] // 2
    lt = s[t[0] - m][t[1] - m]
    lb = s[t[0] + m + 1][t[1] - m]
    rt = s[t[0] - m][t[1] + m + 1]
    rb = s[t[0] + m + 1][t[1] + m + 1]
    print(os - (rb - rt - lb + lt))
# 実行時間 1160ms
